### ISSUE TITLE
Feasibility of periodic boundary conditions?

### ISSUE THREAD

**Issue by ricsonc** [andyljones/pybbfmm#2]
I'm wondering what needs to be changed to be compatible with periodic boundary conditions (my kernel already handles the additional interactions in the infinite "periodic copies" of the environment, so I'm just asking about the FMM part itself). The literature doesn't seem to indicate any reason why it shouldn't be possible (except for stuff like kernels with the form 1/r whose sum doesn't converge, but I don't have that problem.)

The two most glaring things to me seem to be the `neighbour_boxes`, which I modified to allow "wrap around" neighbors by changing `(current >= 0)` to `(current > 0)` in the "ascend to common ancestor". Along with this I also fixed the scaling logic so that the periodic-boundaries get scaled to `[-1,1]^D`, so that what's considered "near" or "far" is properly considered. 

Anyway, I still don't have correct outputs, so I thought I'd ask if there's an obvious reason why what I'm trying to do is impossible, or if I've forgotten to change something really obvious.

**Comment by andyljones** [comment]
From what you've said, my suspicion is that the `current > 0` has some edge-case that's not giving you wrap-around neighbours like you think it does. Or it's giving those neighbours multiple times!

The way to test this is to put together a minimal problem - a couple of sources and targets and `capacity` set to 1 - and take a look at the interaction lists for each box in turn. 

This is a bit tricky in the current version of the code because the lists are only represented implicitly as 'schemes'. In an earlier version of the code they were explicit lists of interacting boxes, and that was easy to plot out with the functions in plotting.py. Each list got its own colour, and if a box had two colours - or, considering the useful-for-debugging `y` list too - no colour, then there was a bug. Kinda like this plot from Carrier:

![image](https://user-images.githubusercontent.com/662606/88758591-b8759280-d160-11ea-8977-bb6b95cdee3e.png)

Put another way, the fundamental invariant is that for any specific box, its partners on the interaction lists `u`, `v`, `w`, `x`, `y` should partition the space. If they don't, you've got a bug.

Unfortunately for performance reasons I replaced all the lists with more loosely-defined 'schemes'. To go from schemes to lists, you'll want to play most of the code from the `_interactions` functions in `__init__.py`. Just rather than compute the `K` of a box and its partners, add that box and its partner to the corresponding list.

**Comment by ricsonc** [comment]
Thanks for the tips. 

I am just confused by one thing -- you said "Or it's giving those neighbours multiple times!" -- but I'm not sure what this means, since the `neighbour_boxes` function returns exactly one neighbour index for each input index. Unless you mean that the same neighbour is returned for two distinct directions, but I thought that was legal -- below is an example from your stack overflow / colab notebook version of `neighbor_boxes` where both the (-1,-1) and also the (0,-1) neighbor of the blue box are the orange box.

To be honest, I didn't find any formal definition on what the neighbours are supposed to be, but from reading the code I assumed it's something like "take the box at the same level/size/depth of the reference box, and in the appropriate direction. if it doesn't exist, take it's parent, and if it's parent doesn't exist, take it's parent's parent, etc"

![image](https://user-images.githubusercontent.com/14364597/88762065-4cc20400-d125-11ea-986e-d7030bce33fb.png)

**Comment by andyljones** [comment]
My bad - it's legal for `neighbour_boxes` to spit it out the same neighbour multiple times, but for the places where I expect that to happen they're then deduplicated. Adding wrapped neighbours might be introducing duplicates where they're not expected, and so (eventually) the same pair of boxes is being fed into `K` more than once. Which is bad. 

(I *think* I mostly rely on `ragged.from_pairs` and `ragged.invert_indices` to do the deduplication, but it's been a long time and I didn't write enough comments).

Neighbours are, yeah, the next box over in a given direction of at least the same size. 

e: To make something clear: 

* Conceptually, I think you're right, that this should generalize to periodic boundaries just fine. 
* I think about the bbfmm as 
  * sources to weights
  * weights upwards into more weights
  * weights downwards into interactions and contributions
  * contributions to targets
* Of those parts, the only bit that should even take notice of periodic boundaries is the interactions.
* *Almost all the bugs I ever had with this project* were from bad interaction lists. 
* That's why I'm homing in on the same thing now.

**Comment by andyljones** [comment]
Ah, there are likely cross-list duplicates. At the least, u and v will clash. In 1D:
```
|       |       |
| b |   |   |   |
```
Then your u-neighbours - the leaves you're next to - are
```
|       |       |
| b | u |   | u |
```
and your v-neighbours - your parent's colleagues' distant children - are
```
|       |       |
| b |   | v | v |
```
Right-hand box shows up in both lists. Ffff.

If I were doing this, I'd pull almost all the logic in the `_interactions` functions back into the `_scheme` functions so that when I did the presolve tree, I ended up with explicit lists of pairs. It's crap from a memory perspective, but it'll make debugging much easier. Then I'd bang my head against the wall getting the lists to partition the space, then  check and recheck it all works on as many edge cases as come to mind, then I'd re-apply the memory optimizations.

**Comment by ricsonc** [comment]
Hm, unless I am misunderstanding something, I think I was able to get the lists out of the scheme without too much trouble:

```
    lists = arrdict.arrdict(
        U = scheme.u.image(box),
        V = torch.cat([sublist.friends[sublist.boxes == box] for sublist in scheme.v]),
        W = scheme.w.image(box),
        X = scheme.x.image(box),
        Y = orthantree.y_pairs(tree, box),
        box = [box],
    )
```

I reproduced the figure from Carrier, and at least for this particular box, the periodic and nonperiodic versions are exactly the same (as expected, since the box is nowhere near a boundary). 

![image](https://user-images.githubusercontent.com/14364597/88880900-cbc94200-d1e2-11ea-9484-bbd2eecfcedb.png)

If I shift the points around (with wraparound), then it's apparent the differences between the original (left) and periodic (right). Actually I don't think there's anything wrong here yet.
![image](https://user-images.githubusercontent.com/14364597/88882104-0f717b00-d1e6-11ea-9634-a5dbc5b48bce.png)


The W-list is incorrect here if I take the bottom right box -- because for the left-colleague of 3, which is 1, the children 5 and 6 are considered distant, whereas the right-colleague of 3 is also 1, and that case considers 7 and 8 to be distant. So it ends up that all 4 are marked as distant, even though none of them are.

(left is original, right with periodic neighbors)
![image](https://user-images.githubusercontent.com/14364597/88881469-59f1f800-d1e4-11ea-95ab-0affa3a4db97.png)

And something similar goes wrong with the V-list here in box 9. 

(left is original, right with periodic neighbors)
![image](https://user-images.githubusercontent.com/14364597/88881570-9d4c6680-d1e4-11ea-88be-f3a9297e45bf.png)

Anyway, I can't think of an easy way off the top of my head to fix this yet, but at least it's pretty clear what's wrong. Thanks for all the pointers!

**Comment by andyljones** [comment]
Well I'll be. I was expecting it to be much more work to get the plotting working again. Great job, and thanks for showing me!

**Comment by ricsonc** [comment]
Sorry to bother you again, but I think I may have spoken a bit too soon about figuring out all the lists. In particular, for the black box below, the box lists don't seem to partition the space (this is with the original, nonperiodic code). I reinstalled a clean copy of the library to make sure I hadn't broken anything while playing around too.

In particular, for the box right below the target box, (labeled 6 in the above images, and 23 being the target box "`b`"), I can't seem to figure out which list it's even *supposed* to go in, according to Carrier. It can't be `U` or `W` because those are empty when b is a parent. It can't be `Y` because it's obviously not well-separated, and it can't be `V` because it's bigger than `b`, *and* not separated. This leaves X, but I don't think that works either, since that also requires separation. I am probably misreading something obvious, but I can't seem to find it.

![image](https://user-images.githubusercontent.com/14364597/89007743-300d0400-d2be-11ea-9081-0022f77cb097.png)

**Comment by ricsonc** [comment]
sorry, wrote a stupid question and then figured it out a minute later.

**Comment by andyljones** [comment]
No worries mate, 's how rubber duck debugging goes 😄 . I've done the exact same plenty of times.

**Comment by ricsonc** [comment]
I've realized I'm still confused about one thing: here I'm taking the black box (23) as `b`. This plot is with a completely fresh copy of the code, so none of the periodic boundary logic. I'm confused about box 6 (not to mention the other uncolored boxes), which doesn't appear to be in any of the lists. I can't even figure out which list it is *supposed* to go in, according to Carrier.

Since `b` is a parent, it can't be in U or W (which are empty). Since `b` and 6 aren't well separated, it can't be V, Y, or X either. I must be misreading something simple in the definitions, but I can't seem to figure it out.

![image](https://user-images.githubusercontent.com/14364597/89009332-305ace80-d2c1-11ea-9874-77482983b37d.png)

**Comment by andyljones** [comment]
Got your comment by email and actually I think this one's my fault. When I said 

> Put another way, the fundamental invariant is that for any specific box, its partners on the interaction lists u, v, w, x, y should partition the space. If they don't, you've got a bug.

I should have said

> Put another way, the fundamental invariant is that for any specific **leaf**, its partners on the interaction lists u, v, w, x, y should partition the space. If they don't, you've got a bug.

In the case of interior boxes, what would be the u- and v- lists of neighbours-and-colleagues-children are in a sense 'too close' to be handled at the level of the box itself, and so instead that part is exactly the part that it's handled by its descendents.

**Comment by andyljones** [comment]
Concretely: 6 will be in lists of each of 50, 52, 49, 51.

**Comment by andyljones** [comment]
Sorry about this, it's been a few months since I wrote pybbfmm and my understanding is obviously starting to decay.

**Comment by ricsonc** [comment]
Thanks!, that clears it up -- the paper itself wasn't clear on this too -- they say the cell is partitioned for any box, but I guess they meant leaf

**Comment by andyljones** [comment]
I *think* - though I'm now uncertain myself - that the union of the u/v/w/x lists over all boxes on a path from the root to a leaf should partition the space. 

Intuition is that for a specific leaf, every other part of the space should be referenced exactly once. By 'referenced', I mean that there's a chain of interaction and inheritance that you can use to get from the leaf to the other part of the space.

The y list for a box is the set of all references that are coming via the box's parent.

**Comment by ricsonc** [comment]
I see, that would make sense, and I believe the `test_lists` function in orthantree is actually checking for that. 

Also, it appears as if i've got everything working now (below figures are the fixed versions of the cross-list duplicates I had before). I only had to tweak W and V slightly

On W, I replaced `distant = (tree.descent[friends] == dirs[:, None, :]).any(-1)`
with a really direct approach of just computing the distances between the boxes and their sizes. Much less elegant, but it seems to work fine.
```
distances = torch.abs(tree.centers[origins].unsqueeze(1) - tree.centers[friends])
distances = torch.min(distances, 2-distances) 
sum_radii = 0.5**tree.depths[origins].float().unsqueeze(1) + 0.5**tree.depths[friends].float()
distant = (sum_radii.unsqueeze(-1) < distances).max(axis = -1).values
```
On V, within the inner-most loop over depth, I added
```
period = 2**depth
periodic_offset = (offset + period // 2) % period - (period // 2)
if (periodic_offset.abs() <= 1).all(-1):
    continue
```
just to check for separation in both directions.

I also needed to remove duplicates, since in some cases a box in the V or W list could be reached either by going through the left neighbor or the right neighbor, I think you already mentioned something to this effect. But since both V and W deal with colleagues, this wrap-around is only a problem at depth 1 (since at depth >= 2, there's no possibility of a duplicated neighbor), so I simply restricted the neighbors at depth 1 to the non-negative quadrant (all depth-1 neighbors are always accessible exactly once through the nonnegative quadrant when there is wraparound). This seems to eliminate all the duplicates. 

I generated some test problems and compared with a brute-force solution, and everything seems to work out, so I think we can declare victory :)

![image](https://user-images.githubusercontent.com/14364597/89011941-26879a00-d2c6-11ea-968d-86a73c95a621.png)
![image](https://user-images.githubusercontent.com/14364597/89011959-2e473e80-d2c6-11ea-98d0-6d62ca84e985.png)
